proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,587 @@
+- Download and run [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip) on that IntellIJ instance.
+
+## How to try KEEP-87? (Alternative approach)
+
+- Clone [Our Kotlin fork](https://github.com/arrow-kt/kotlin) and checkout the **keep-87** branch.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#build-environment-requirements) configure the necessary JVMs.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#-working-with-the-project-in-intellij-idea) to open the project in IntelliJ IDEA.
+- Once you have everything working, you can run a new instance of IntelliJ IDEA with the new modifications to the language by executing `./gradlew runIde`. There is also a pre-configured run configuration titled **IDEA** that does this.
+- It will open a new instance of the IDE where you can create a new project and experiment with the new features of the language. You can also download [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip)
+) with some sample code that you can try out.
+
+
+## What's still to be done?
+
+### Instance resolution based on inheritance
+
+Some scenarios are not covered yet given some knowledge we are lacking about how subtyping resolution rules are coded in Kotlin compiler. The different scenarios would be required for a fully working compile time extension resolution feature, and [are described in detail here](https://github.com/arrow-kt/kotlin/issues/15).
+
+### Using extensions in inlined lambdas
+
+Inlined functions get into trouble when it comes to capture resolved extensions.[The problem is described here](https://github.com/arrow-kt/kotlin/issues/14).
+
+### Function and property extension providers
+
+Ideally we'd enable users to provide extensions also using `val` and `fun`. They'd look similar to:
+
+```kotlin
+// Simple fun extension provisioning
+extension fun userRepository(): Repository<User> = object : Repository<User>() {
+	/* ... */
+}
+
+// Chained fun extension provisioning (would require both to resolve).
+extension fun userValidator(): Validator<User> = UserValidator()
+extension fun userRepository(with validator: Validator<User>) : Repository<User> = UserRepository(validator)
+
+// Simple extension provisioning
+extension val userRepository: Repository<User> = UserRepository()
+
+// Chained extension provisioning (would require both to resolve).
+extension val userValidator: Validator<User> = UserValidator()
+extension val userRepository: Repository<User> = UserRepository(userValidator)
+```
+
+In chained extension provisioning, all chained extensions would be required from the caller scope, but not necessarily all provided in the same resolution scope. E.g: `Repository<User>` could be provided in a different resolution scope than `Validator<User>`, and the program would still compile successfully as long as both are available.
```

Comment:
I might not be using the proper term here (open for suggestions) but I'm referring to nesting extension resolutions up to N arbitrary levels. An extension can require a constraint that is provided somewhere else (e.g: Dagger providers). That's something already working for classes and objects, so we would also be needed for funs and vals. Just one pass is needed AFAIK.

About "could be provided in a different scope" I mean you can compose dependency resolution from N different resolution scopes as in the ordered steps (scopes) we look at for resolution described on this keep. On this example, `Repository<User>` could be defined in the target type companion and `Validator<User>` in the type class subpackages (just a random example), and it should suffice, since both are properly resolved.

Reply:
Right, switched to `nested` in 83a1a94.